pkg/gooq/select.go

```diff
@@ -265,9 +272,42 @@ func (s *selection) Render(
 	}
 
 	// render WHERE clause
-	if len(s.predicate) > 0 {
+
+	predicate := s.predicate
+	if len(s.seek) > 0 {
+		// faster and stable pagination based on these two articles
+		// https://blog.jooq.org/2013/10/26/faster-sql-paging-with-jooq-using-the-seek-method/
+		// https://blog.jooq.org/2013/11/18/faster-sql-pagination-with-keysets-continued/
+		numOrdering := len(s.ordering)
+		for index, order := range s.ordering {
+			isLast := index == numOrdering-1
+			var operand Expression
+			var operator Operator
+			switch order.getOperator() {
+			case OperatorDesc:
+				operand = order.getExpressions()[0]
+				operator = OperatorLt
+			case OperatorAsc:
+				operand = order.getExpressions()[0]
+				operator = OperatorGt
+			default:
+				operand = order
+				operator = OperatorGt
+			}
+			if !isLast && operator == OperatorGt {
```

Comment:
I am not sure if it makes sense to use `>=` for all of the conditions before the last condition and use `>` for the last condition.

Reply:
I agree that this logic doesn't seem to match the algorithm outlined in the article you linked.

Here's a test that highlights this PR's logic:

```go
	{
		Constructed:  Select().From(Table1).OrderBy(Table1.Column1.Desc(), Table1.ID.Desc()).Seek("foo", "bar"),
		ExpectedStmt: `SELECT * FROM public.table1 WHERE table1.column1 <= $1 AND table1.id < $2 ORDER BY table1.column1 DESC, table1.id DESC`,
		Arguments:    []interface{}{"foo", "bar"},
	},
```

Note that `table1.id < $2` **must** be `true` for a row to be included in the result.

In jOOQ, the generated statement would look something like this:

```sql
SELECT *
FROM public.table1
WHERE table1.column1 < $1 OR (table1.column1 = $1 AND table1.id < $2)
ORDER BY table1.column1 DESC, table1.id DESC
```

In other words, `table1.id < $2` must only be true if there is a "tie" on the first argument.

See the example [here](https://www.jooq.org/doc/latest/manual/sql-building/sql-statements/select-statement/seek-clause/). Note that `(37, 3)` is included in the result, even though `37 < 533`. The row is included because `3 > 2`.

Reply:
That's a good point. Let me revise the generated statement.

Reply:
I am wondering if given the following scenario
```
Select().From(Table1).OrderBy(Table1.Column1.Desc(), Table1.Column2.Desc(), Table1.Column3.Desc()).Seek("foo1", "foo2", "foo3"),
```
would the follow where clause be correct
```
WHERE ((column1 < "foo1")
  OR (value1 = "foo1" AND value2 < "foo2")
  OR (value1 = "foo1" AND value2 = "foo2" AND value3 < "foo3"))
```

Reply:
This look consistent with https://github.com/jOOQ/jOOQ/blob/94d850789f62e90b207e195555e292567bc0fb1e/jOOQ/src/main/java/org/jooq/impl/SelectQueryImpl.java#L2241

Reply:
I haven't seen a jOOQ example with `3 >=` arguments, but I suspect that is the desired behavior; subsequent arguments are only observed in the case of a tie.

Reply:
Should be fixed now